### PR TITLE
Description Lists Extension

### LINKED ISSUE #52: Description List as an extension
There is an open [thread about description lists](https://talk.commonmark.org/t/description-list/289) in [talk.commonmark.org](https://talk.commonmark.org/). Since comrak has support for [some extensions](https://github.com/kivikakk/comrak/blob/dfe55ae29482c038c0ba2f4adf925421c6709008/src/parser/mod.rs#L136-L254), I think that it would be nice to add support for definition lists.

## Syntax

There is no official syntax for definition list, but it is implemented in many Markdown processors. One of the comments includes [a list of lins](https://talk.commonmark.org/t/description-list/289/11) to the documentation of some of those processors.

The most common syntax is using colon at the beginning of the definition.

    term1

    : definition1

    term2

    : definition2

This syntax is used in [Pandoc](https://pandoc.org/MANUAL.html#definition-lists), [PHP Markdown](https://michelf.ca/projects/php-markdown/extra/#def-list), [fletcher/MultiMarkdown-4](https://fletcher.github.io/MultiMarkdown-4/definitionlists), [Python Markdown](https://python-markdown.github.io/extensions/definition_lists/), [kramdown](https://kramdown.gettalong.org/quickref.html#definition-lists), etc.

### PR DESCRIPTION
**Pull request by ayosec**
This patch adds support for description lists, as described in #52.

## Overview

A description-list is written as a group of terms, each one with a description (*details*). It is required to put a blank line between terms and descriptions.

For example, the following source:

```
Term 0

: Description 0

Term 1

: Description 1

    More info
```

Is rendered to:

```html
<dl>
  <dt>
    <p>Term 0</p>
  </dt>
  <dd>
    <p>Description 0</p>
  </dd>
  <dt>
    <p>Term 1</p>
  </dt>
  <dd>
    <p>Description 1</p>
    <p>More info</p>
  </dd>
</dl>
```

It also supports nested nodes. The following source:

```
0

1

: 2

3

: 4

    1. 5

    2. 6
        > 7
        >
        > : 8
        >
        > 9
        >
        > : 10
        >
        > 11
```

Is rendered to:

```html
<p>0</p>
<dl>
  <dt>
    <p>1</p>
  </dt>
  <dd>
    <p>2</p>
  </dd>
  <dt>
    <p>3</p>
  </dt>
  <dd>
    <p>4</p>
    <ol>
      <li>
        <p>5</p>
      </li>
      <li>
        <p>6</p>
        <blockquote>
          <dl>
            <dt>
              <p>7</p>
            </dt>
            <dd>
              <p>8</p>
            </dd>
            <dt>
              <p>9</p>
            </dt>
            <dd>
              <p>10</p>
            </dd>
          </dl>
          <p>11</p>
        </blockquote>
      </li>
    </ol>
  </dd>
</dl>
```

## Implementation Details

I tried to imitate the support for regular lists in some parts of the implementation, mostly to support nested nodes.

There are four new node variants in `NodeValue`:

* `DescriptionList`
* `DescriptionItem`
* `DescriptionTerm`
* `DescriptionDetails`

`DescriptionItem` is used only for the parser. It is not emitted in the HTML.

To detect a description list:

* The parser checks for a `:` symbol as the first non-space character.
* When it is found, it checks if the previous node in the same container is a paragraph .
* If it is paragraph:
    * Detach it
    * Check if the (new) last child is a `DescriptionList`.
        * If not, add a new `DescriptionList` node.
    * Creates a tree like `DescriptionItem / DescriptionTerm / $old-paragraph`, and append it to the `DescriptionList` node.
    * Finally, create a `DescriptionDetails` node under `DescriptionItem`, and use it as the new container.

A tricky part of the implementation is that I have to change the `open` field of an existing `DescriptionList`. This is required because every new term is created as a paragraph (before detect the `:` line with the description), so the previous list is closed. I don't know if this is the proper way to do it.

To visualize this:

If we have this source:

```
A

: B

C

: D
```

The behaviour of the parser is something like:

* Found `A`:
    * Create a `Paragraph` node, and append it to the top container.
    * The document is:
        ```
        Document
            Paragraph "A"
        ```
* Found `:`:
    * Detects that last child is a `Paragraph`, so it is detached.
    * There is no `DescriptionList` at the end of the container, so it creates a new one.
    * Append the detached paragraph to the new list.
    * The document is:
        ```
        Document
            DescriptionList
                DescriptionItem
                    DescriptionTerm
                        Paragraph "A"
                    DescriptionDetails
        ```
    * `DescriptionDetails` is new current container to add nodes.
* Found `B` after `:`:
    * Creates a `Paragraph` and add it to the `DescriptionDetails` node.
* Found `C`:
    * Closes the current `DescriptionList` and add the text as a new `Paragraph`.
    * The document is:
        ```
        Document
            DescriptionList
                DescriptionItem
                    DescriptionTerm
                        Paragraph "A"
                    DescriptionDetails
                        Paragraph "B"
            Paragraph "C"
        ```
* Found `:`:
    * Detects that last child is a `Paragraph`, so it is detached.
    * After detach, found a `DescriptionList` as the last child.
        * Mark its nodes as `open`.
    * Append the `DescriptionItem` node to the current list.
    * The document is:
        ```
        Document
            DescriptionList
                DescriptionItem
                    DescriptionTerm
                        Paragraph "A"
                    DescriptionDetails
                        Paragraph "B"
                DescriptionItem
                    DescriptionTerm
                        Paragraph "C"
                    DescriptionDetails
        ```

The requirement of the blank line between terms and descriptions is to keep the implementation simpler. Many parsers supports a syntax like:

```
term 1
: description 1
term 2
: description 2
```

But it was not clear to me how to support this. I guess that this is something that can be added later.

## Changes in the CLI

The CLI supports the flag `description-lists` to enable the extension.

## S-Expressions in the examples

An example program in `examples/s-expr.rs` can be used to print the AST as S-expressions.

```bash
$ cargo run --example s-expr file1.md file2.md ...
$ cat file.md | cargo run --example s-expr 
```

For example, the following source:

```markdown
Lorem ipsum dolor sit amet:

* Consectetur adipiscing elit
* Sed do **eiusmod** tempor incididunt
    * Ut labore et [dolore](magna-aliqua).
    * Ut enim ad minim veniam.
```

Generates this output:

```
$ cargo run --example s-expr path/to/file.md
(Document
    (Paragraph "Lorem ipsum dolor sit amet.")
    (List(NodeList { list_type: Bullet, marker_offset: 0, padding: 2, start: 1, delimiter: Period, bullet_char: 42, tight: true })
        (Item(NodeList { list_type: Bullet, marker_offset: 0, padding: 2, start: 1, delimiter: Period, bullet_char: 42, tight: false })
            (Paragraph "Consectetur adipiscing elit"))
        (Item(NodeList { list_type: Bullet, marker_offset: 0, padding: 2, start: 1, delimiter: Period, bullet_char: 42, tight: false })
            (Paragraph "Sed do " (Strong "eiusmod") " tempor incididunt")
            (List(NodeList { list_type: Bullet, marker_offset: 2, padding: 2, start: 1, delimiter: Period, bullet_char: 42, tight: true })
                (Item(NodeList { list_type: Bullet, marker_offset: 2, padding: 2, start: 1, delimiter: Period, bullet_char: 42, tight: false })
                    (Paragraph "Ut labore et " (Link(NodeLink { url: [109, 97, 103, 110, 97, 45, 97, 108, 105, 113, 117, 97], title: [] }) "dolore") "."))
                (Item(NodeList { list_type: Bullet, marker_offset: 2, padding: 2, start: 1, delimiter: Period, bullet_char: 42, tight: false })
                    (Paragraph "Ut enim ad minim veniam.")))))
)
```

I added it to the patch because it was useful to debug the AST, and it may be useful for other people. If you think that it should not be included, I will remove it.

----

Closes #52

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -64,6 +64,7 @@ fn main() {
                     "tasklist",
                     "superscript",
                     "footnotes",
+                    "description-lists",
```
Comment (kivikakk) [inline src/main.rs]: We add `description-lists` as a possible `-e` value here …

File: src/main.rs [hunk 2]
```diff
@@ -98,6 +99,11 @@ fn main() {
                 .long("footnotes")
                 .help("Parse footnotes"),
         )
+        .arg(
+            clap::Arg::with_name("description-lists")
+                .long("description-lists")
+                .help("Parse description lists"),
+        )
```
Comment (kivikakk) [inline src/main.rs]: … and add `--description-lists` as a separate flag here ...

File: src/main.rs [hunk 3]
```diff
@@ -125,6 +131,7 @@ fn main() {
         ext_superscript: exts.remove("superscript"),
         ext_header_ids: matches.value_of("header-ids").map(|s| s.to_string()),
         ext_footnotes: matches.is_present("footnotes"),
+        ext_description_lists: matches.is_present("description-lists"),
```
Comment (kivikakk) [inline src/main.rs]: … and use the flag toggle here, not the `exts` set. As a result:

```console
$ cargo run <(echo "a"; echo; echo ": b") --description-lists
    Finished dev [unoptimized + debuginfo] target(s) in 0.06s
     Running `target/debug/comrak /dev/fd/63 --description-lists`
<dl><dt>
<p>a</p>
</dt>
<dd>
<p>b</p>
</dd>
</dl>
$ cargo run <(echo "a"; echo; echo ": b") -e description-lists
    Finished dev [unoptimized + debuginfo] target(s) in 0.06s
     Running `target/debug/comrak /dev/fd/63 -e description-lists`
thread 'main' panicked at 'assertion failed: exts.is_empty()', src/main.rs:137:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.
$
```

Ideally description lists should be an `-e` option. (And so should footnotes, frankly.)

Comment (ayosec) [reply src/main.rs]: Oops.. I didn't notice that =/.

Fixed in d7469e70d790ca5619faf5d4fe6f77a7a17a4c24.

**Comment by kivikakk** [comment]
Thank you! ❤️

**Comment by brson** [comment]
Awesome writeup and patch.

**Pull request merged**